lib/src/rules/avoid_leading_underscore_for_non_private_identifiers.dart

```diff
@@ -0,0 +1,96 @@
+// Copyright (c) 2015, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+import 'package:linter/src/analyzer.dart';
+import 'package:linter/src/utils.dart';
+
+const _desc = r'Avoid leading underscore for identifiers that aren’t private.';
```

Comment:
Both of these are more food for thought than a request for a change.

It isn't so much that they "aren't" private, but that they "can't be" private (or conversely that they can't be public). Maybe a better way to say it is "local" identifiers.

Along those lines, the name of the lint is kind of long. Some of the other lint rules we have that are related to naming conventions have names like "file_names", "library_names", "package_names", "constant_identifier_names", and "non_constant_identifier_names". Perhaps for consistency something like "local_variable_names"? The disadvantage of this is that "constant_identifier_names" and "non_constant_identifier_names" already check local variables (but not for leading underscores), so it might be somewhat confusing. No concrete suggestion, just a notion that a shorter name would be better.

Reply:
I also had the same confusion. Initially I wanted to name like "avoid_leading_underscore_for_local_identifiers". But this rule also check for library prefixes, which I am not sure whether categorised as local identifier.

How about we use this rule only for local identifiers and move library prefix check to existing "library_prefixes" rule? Or shall we create new one like "avoid_leading_underscore_for_library_prefixes".

Reply:
> But this rule also check for library prefixes, which I am not sure whether categorised as local identifier.

No, we generally don't consider a library prefix to be a local variable. We generally use "local" to mean local to a function / method.

Perhaps we should flip the sense of the name and call it 'private_names' and describe it as a rule to ensure that underscore is only used for identifiers that would be public if the underscore were absent.

> How about we ... move library prefix check to existing "library_prefixes" rule?

I considered that possibility. We could also move the local variable checking into "constant_identifier_names" and "non_constant_identifier_names".

The problem is that changing any of these three rules is potentially a breaking change. If we have users that have enabled those rules but currently have local variables with underscores, then when they upgrade their SDK they'll start seeing lints being reported that might cause their builds to stop working. We generally avoid making breaking changes like that (we can't always, such as for some bug fixes, but we try).

Similarly, if some users intentionally use underscores for some reason, then that makes these rules unusable for them.

I think overall it's better to create a new rule the way you did than to risk breaking existing users.

Reply:
I realised that `non_private_identifiers` is not the right term to use here. And updating existing rules is also not a good idea. So I come up with two lint rules,

- avoid leading underscore for local identifiers
- avoid leading underscore for library prefixes

I updated the PR accordingly. Let me know if you have any suggestions.

Reply:
Thanks for the thoughtful follow-up!

I feel inclined to focus on avoiding leading underscores for local identifiers here and then investigate updating `library_prefixes` to check for underscores.  Before we do that though I'd want to do a bit of investigating to see if it's really what we want to do.  (For example, is this something that we do internally?  cc @srawlins who may know....)

@jawahars16: have you seen this naming in practice?